go-controller/pkg/ovn/egressip.go

```diff
@@ -0,0 +1,542 @@
+		if node, exists := oc.eIPAllocator[status.Node]; exists {
+			delete(node.allocations, status.EgressIP)
+		}
+		klog.V(5).Infof("Remaining allocations on node are: %+v", oc.eIPAllocator[status.Node])
+	}
+	return nil
+}
+
+func (oc *Controller) getSortedEgressData() ([]eNode, map[string]bool) {
+	nodes := []eNode{}
+	allAllocations := make(map[string]bool)
+	for _, eNode := range oc.eIPAllocator {
+		nodes = append(nodes, *eNode)
+		for ip := range eNode.allocations {
+			allAllocations[ip] = true
+		}
+	}
+	sort.Slice(nodes, func(i, j int) bool {
+		return len(nodes[i].allocations) < len(nodes[j].allocations)
+	})
+	return nodes, allAllocations
+}
+
+func getEgressIPKey(eIP *egressipv1.EgressIP) string {
+	return eIP.Name
+}
+
+func getPodKey(pod *kapi.Pod) string {
+	return fmt.Sprintf("%s_%s", pod.Namespace, pod.Name)
+}
+
+type egressIPMode struct {
+	ovnNBClient    goovn.Client
+	podRetry       sync.Map
+	gatewayIPCache sync.Map
+}
+
+type gatewayRouter struct {
+	ipV4 net.IP
+	ipV6 net.IP
+}
+
+func (e *egressIPMode) getGatewayRouterIP(node string, wantsIPv6 bool) (net.IP, error) {
+	if item, exists := e.gatewayIPCache.Load(node); exists {
+		if gatewayRouter, ok := item.(gatewayRouter); ok {
+			if wantsIPv6 && gatewayRouter.ipV6 != nil {
+				return gatewayRouter.ipV6, nil
+			} else if !wantsIPv6 && gatewayRouter.ipV4 != nil {
+				return gatewayRouter.ipV4, nil
+			}
+			return nil, fmt.Errorf("node does not have a gateway router IP corresponding to the IP version requested")
+		}
+		return nil, fmt.Errorf("unable to cast node: %s gatewayIP cache item to correct type", node)
+	}
+	gatewayRouter := gatewayRouter{}
+	err := utilwait.ExponentialBackoff(retry.DefaultRetry, func() (bool, error) {
+		grIPv4, grIPv6, err := util.GetNodeGatewayRouterAddress(node)
+		if err != nil {
+			return false, nil
+		}
+		if grIPv4 != nil {
+			gatewayRouter.ipV4 = grIPv4
+		}
+		if grIPv6 != nil {
+			gatewayRouter.ipV6 = grIPv6
+		}
+		return true, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	e.gatewayIPCache.Store(node, gatewayRouter)
+	if wantsIPv6 && gatewayRouter.ipV6 != nil {
+		return gatewayRouter.ipV6, nil
+	} else if !wantsIPv6 && gatewayRouter.ipV4 != nil {
+		return gatewayRouter.ipV4, nil
+	}
+	return nil, fmt.Errorf("node does not have a gateway router IP corresponding to the IP version requested")
+}
+
+func (e *egressIPMode) getPodIPs(pod *kapi.Pod) ([]net.IP, error) {
+	_, podIps, err := util.GetPortAddresses(getPodKey(pod), e.ovnNBClient)
+	if err != nil || len(podIps) == 0 {
+		e.podRetry.Store(getPodKey(pod), true)
+		return nil, fmt.Errorf("pod does not have IP assigned, err: %v", err)
+	}
+	if e.needsRetry(pod) {
+		e.podRetry.Delete(getPodKey(pod))
+	}
+	return podIps, nil
+}
+
+func (e *egressIPMode) needsRetry(pod *kapi.Pod) bool {
+	_, retry := e.podRetry.Load(getPodKey(pod))
+	return retry
+}
+
+func (e *egressIPMode) createEgressPolicy(podIps []net.IP, status egressipv1.EgressIPStatus, mark uint32) error {
```

Comment:
you should have some documentation on this function and explain what mark is (i'm assuming the packet mark)

Reply:
Yes, packet mark. I'll add some documentation